File: 1335-minimum-difficulty-of-a-job-schedule/1335-minimum-difficulty-of-a-job-schedule.cpp
class Solution {
public:
    int dp[301][11];
    int minDifficulty(vector<int>& jobDifficulty, int d) {
        int n = jobDifficulty.size();
        memset(dp, -1, sizeof(dp));
        if(n < d) return -1;
        return f(jobDifficulty, n, 0, d);
    }
    int f(vector<int> &jobD, int n, int i, int d){
        if(d == 1) return *max_element(jobD.begin()+i, jobD.end());
        if(dp[i][d] != -1) return dp[i][d];
        int ans = INT_MAX, curr = INT_MIN;
        for(int j=i; j<=n-d; j++){
            curr = max(curr, jobD[j]);
            ans = min(ans, curr + f(jobD, n, j+1, d-1));
        }
        return dp[i][d] = ans;
    }
};

File: 1335-minimum-difficulty-of-a-job-schedule/1335-minimum-difficulty-of-a-job-schedule.java
class Solution {
    int[][] dp = new int[301][11];
    public int minDifficulty(int[] jobDifficulty, int d) {
        int n = jobDifficulty.length;
        for(int[] x : dp) Arrays.fill(x, -1);
        if(n < d) return -1;
        return f(jobDifficulty, n, 0, d);
    }
    int f(int[] jobD, int n, int i, int d){
        if(d == 1){
            int mx = Integer.MIN_VALUE;
            for(int j=i; j<n; j++) mx = Math.max(mx, jobD[j]);
            return mx;
        }
        if(dp[i][d] != -1) return dp[i][d];
        int ans = Integer.MAX_VALUE, curr = Integer.MIN_VALUE;
        for(int j=i; j<=n-d; j++){
            curr = Math.max(curr, jobD[j]);
            ans = Math.min(ans, curr + f(jobD, n, j+1, d-1));
        }
        dp[i][d] = ans;
        return ans;
    }
}